pkg/volume/empty_dir/empty_dir_linux.go

```diff
@@ -35,18 +36,33 @@ type realMountDetector struct {
 	mounter mount.Interface
 }
 
-func (m *realMountDetector) GetMountMedium(path string) (storageMedium, bool, error) {
-	glog.V(5).Infof("Determining mount medium of %v", path)
-	notMnt, err := m.mounter.IsLikelyNotMountPoint(path)
+func (m *realMountDetector) GetMountMedium(p string) (storageMedium, bool, error) {
+	glog.V(5).Infof("Determining mount medium of %v", p)
+	notMnt, err := m.mounter.IsLikelyNotMountPoint(p)
 	if err != nil {
-		return 0, false, fmt.Errorf("IsLikelyNotMountPoint(%q): %v", path, err)
+		return 0, false, fmt.Errorf("IsLikelyNotMountPoint(%q): %v", p, err)
 	}
+
+	// Workaround buggy filesystems like vmhgfs which fail on statfs with long
+	// file names, walk up the tree stopping just in front of the filesystem border
+	stat := syscall.Stat_t{}
+	if err = syscall.Stat(p, &stat); err != nil {
+		return 0, false, fmt.Errorf("stat(%q): %v", p, err)
+	}
+	dev := stat.Dev
+	dotdot := p
+	for dotdot != "/" && err == nil && dev == stat.Dev {
+		p, dotdot = dotdot, path.Dir(p)
+		dev = stat.Dev
+		err = syscall.Stat(dotdot, &stat)
+	}
+
 	buf := syscall.Statfs_t{}
-	if err := syscall.Statfs(path, &buf); err != nil {
-		return 0, false, fmt.Errorf("statfs(%q): %v", path, err)
+	if err := syscall.Statfs(p, &buf); err != nil {
```

Comment:
nothing actually guarantees this will work, right?  If a filesystem was mounted at a long path, your climbing will not go far and this will still break.  What do we do then?